pkg/operator/configobservation/apiserver/observe_apiserver.go

```diff
@@ -0,0 +1,289 @@
+
+	apiServer, err := listers.APIServerLister.Get("cluster")
+	if errors.IsNotFound(err) {
+		// no error, just clear the observed config and observed resources
+		glog.Warningf("apiservers.%s/cluster: %v", configv1.GroupName, err)
+		return nil, append(errs, syncObservedResources(resourceSync, deleteSyncRules(o.resourceNames...))...)
+	}
+
+	// if something went wrong, keep the previously observed config and resources
+	if err != nil {
+		glog.Warningf("error getting apiservers.%s/cluster: %v", configv1.GroupName, err)
+		return previouslyObservedConfig, append(errs, err)
+	}
+
+	observedConfig, observedResources, errs := o.observerFunc(apiServer, recorder, previouslyObservedConfig)
+
+	// if we get error during observation, skip the merging and return previous config and errors.
+	if len(errs) > 0 {
+		glog.Warningf("errors during apiservers.%s/cluster processing: %v", configv1.GroupName, errs)
+		return previouslyObservedConfig, append(errs, errs...)
+	}
+
+	// default to deleting previous resources, and then merge in observed resources rules
+	resourceSyncRules := deleteSyncRules(o.resourceNames...)
+	if err := mergo.Merge(&resourceSyncRules, &observedResources, mergo.WithOverride); err != nil {
+		glog.Warningf("merging resource sync rules failed: %v", err)
+	}
+
+	errs = append(errs, syncObservedResources(resourceSync, resourceSyncRules)...)
+
+	return observedConfig, errs
+}
+
+// deleteSyncRules generates resource sync rules to delete the resources, specified by names, from the
+// operator namespace.
+func deleteSyncRules(names ...string) syncActionRules {
+	resourceSyncRules := syncActionRules{}
+	for _, name := range names {
+		// empty string here means there is "from" anymore and we mark the "name" for deletion.
+		resourceSyncRules[name] = ""
+	}
+	return resourceSyncRules
+}
+
+// syncObservedResources synchronizes resource from the global user specified config namespace into the operator
+// namespace. The the entry keys in the syncRules should be the names of the resources to be synced into the operator
+// namespace. The corresponding values should be either the name of the resource to copy from the global user specified
+// config namespace to the operator's namespace on sync, an empty string to indicate that the resource should be deleted
+// on sync, or a nil to indicate that the resource should be left as-is on sync.
+func syncObservedResources(syncResource resourceSyncFunc, syncRules syncActionRules) []error {
+	var errs []error
+	for to, from := range syncRules {
+		var source resourcesynccontroller.ResourceLocation
+		if len(from) > 0 {
+			source = resourcesynccontroller.ResourceLocation{Namespace: operatorclient.GlobalUserSpecifiedConfigNamespace, Name: from}
+		}
+		// if 'from' is empty, then it means we want to delete
+		destination := resourcesynccontroller.ResourceLocation{Namespace: operatorclient.OperatorNamespace, Name: to}
+		err := syncResource(destination, source)
+		if errors.IsNotFound(err) {
+			continue
```

Comment:
does this mean we ignore missing sources? That's will result in pods not starting as we will mount mount those secrets/configmaps.

Reply:
@sttts you asked to make sure we skip missing :-)

Reply:
we don't have to handle this, it is handled in sync functions, dropping.